Mu2eG4/src/GammaConvGenerator_module.cc

```diff
@@ -0,0 +1,214 @@
+    const auto gammapart = gammas.at(eng_.operator unsigned int() % gammas.size());
+    const auto p4_gamma = gammapart->endMomentum();
+    const auto t_gammastop = gammapart->endGlobalTime();
+    if (verbosity_>2){
+      std::cout << "GAMMA 4-MOMENTUM: " << p4_gamma << std::endl;
+      std::cout << "END VOLUME INDEX: " << gammapart->endVolumeIndex() << std::endl;
+    }
+    //
+    // Handle to information about G4 physical volumes.
+    bool PhysicalVolumeInfoFound=false;
+    std::string wantedbranch="mu2e::PhysicalVolumeInfomvs_"+physVolModuleLabel_+"_eventlevel_"+physVolProcessLabel_+".";
+    if (verbosity_>2) std::cout << "Searching for " << wantedbranch << std::endl;
+
+    std::vector< art::Handle<PhysicalVolumeInfoMultiCollection> > vah = event.getMany<PhysicalVolumeInfoMultiCollection>();
+    art::Handle<PhysicalVolumeInfoMultiCollection> PhysicalVolumeInfos;
+
+    for (auto const & ah : vah){
+      std::string branchName = ah.provenance()->productDescription().branchName();
+      if (verbosity_>2) std::cout << "PHYSICALVOLUMEINFO COLLECTION: " << branchName << std::endl;
+      if (branchName==wantedbranch){
+        PhysicalVolumeInfoFound=true;
+        PhysicalVolumeInfos=ah;
+        if (verbosity_>2){
+          std::cout << "PHYSICALVOLUMEINFO COLLECTION FOUND" << std::endl;
+          std::cout << PhysicalVolumeInfos << std::endl;
+        }
+      }
+    }
+    if (!PhysicalVolumeInfoFound){
+      throw   cet::exception("NO PHYSICALVOLUMEINFO COLLECTION")
+        <<"GammaConvGenerator::produce(): no suitable PhysicalVolume collection\n";
+    }
+    if (PhysicalVolumeInfos->empty()){
+      throw   cet::exception("EMPTY PHYSICALVOLUMEINFO COLLECTION")
+        <<"GammaConvGenerator::produce(): PhysicalVolume collection is empty\n";
+    }
+    //const auto& vols = PhysicalVolumeInfos->at(gammapart->simStage());
+    /*
+    PhysicalVolumeInfoSingleStage const* vols = &PhysicalVolumeInfos->at(gammapart->simStage());
+    if (verbosity_>2){
+      std::cout << "VOLUME COLLECTION SIZE: " << vols->size() << std::endl;
+      for ( long unsigned int i=0;i<vols->size();i++){
+        std::cout << i << " "  << vols->at(cet::map_vector_key(i)).name() << " made of " << vols->at(cet::map_vector_key(i)).materialName() << std::endl;
+        //for (auto const & vol : vols){
+        //	std::cout << vol << " "  << vol.name() << " made of " << vol.materialName() << std::endl;
+      }
+    }
+    //const auto& endInfo = vols.at(gammapart->endVolumeIndex());
+    PhysicalVolumeInfo const& endInfo = vols->at(cet::map_vector_key(gammapart->endVolumeIndex()));
+
+    if (verbosity_>1){
+      std::cout << "Gamma stop found in: " << endInfo.materialName() << std::endl;
+    std::cout << "Requested material is:" << " " << stopMaterial_ << std::endl;
+    }
+    G4Material*stopMat = findMaterialOrThrow((G4String const& )endInfo.materialName());
+    */
+    G4Material*stopMat = findMaterialOrThrow((G4String const& )stopMaterial_);
```

Comment:
I am 99% sure that geometry and materials are defined at beginRun time.  So you can move the lookup of the element fractions to beginRun and save only the information you need.

Reply:
geometry readout moved to beginrun

Reply:
unnecessary PhysVol collection unpacking deleted